wgengine/userspace.go

```diff
@@ -360,22 +368,50 @@ func (e *userspaceEngine) isLocalAddr(ip packet.IP) bool {
 
 // handleDNS is an outbound pre-filter resolving Tailscale domains.
 func (e *userspaceEngine) handleDNS(p *packet.ParsedPacket, t *tstun.TUN) filter.Response {
-	if e.resolver.AcceptsPacket(p) {
-		// TODO(dmytro): avoid this allocation without having tsdns know tstun quirks.
-		buf := make([]byte, tstun.MaxPacketSize)
-		offset := tstun.PacketStartOffset
-		response, err := e.resolver.Respond(p, buf[offset:])
-		if err != nil {
-			e.logf("DNS resolver error: %v", err)
-		} else {
-			t.InjectInboundDirect(buf[:offset+len(response)], offset)
+	if p.DstIP == magicDNSIP && p.DstPort == magicDNSPort && p.IPProto == packet.UDP {
+		request := tsdns.Packet{
+			SrcIP:   uint32(p.SrcIP),
```

Comment:
ah, from here. I don't think we need to spread this further. That'll make fixing ParsedPacket easier later.